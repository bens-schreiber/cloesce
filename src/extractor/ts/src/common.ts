export type Either<L, R> = { ok: false; value: L } | { ok: true; value: R };
export function left<L>(value: L): Either<L, never> {
  return { ok: false, value };
}
export function right<R>(value: R): Either<never, R> {
  return { ok: true, value };
}

export enum ExtractorErrorCode {
  UnknownType,
  MultipleGenericType,
  InvalidIncludeTree,
  UnknownNavigationPropertyReference,
  InvalidNavigationPropertyReference,
  MissingNavigationPropertyReference,
  MissingManyToManyUniqueId,
  MissingPrimaryKey,
  MissingWranglerEnv,
  TooManyWranglerEnvs,
  TooManyMiddlewares,
  MissingMiddlewareMethod,
}

const errorInfoMap: Record<
  ExtractorErrorCode,
  { description: string; suggestion: string }
> = {
  [ExtractorErrorCode.UnknownType]: {
    description: "Encountered an unknown or unsupported type",
    suggestion: "Refer to the documentation on valid Cloesce TS types",
  },
  [ExtractorErrorCode.MultipleGenericType]: {
    description: "Cloesce does not yet support types with multiple generics",
    suggestion:
      "Simplify your type to use only a single generic parameter, ie Foo<T>",
  },
  [ExtractorErrorCode.InvalidIncludeTree]: {
    description: "Invalid Include Tree",
    suggestion:
      "Include trees must only contain references to a model's navigation properties.",
  },
  [ExtractorErrorCode.UnknownNavigationPropertyReference]: {
    description: "Unknown Navigation Property Reference",
    suggestion:
      "Verify that the navigation property reference model exists, or create a model.",
  },
  [ExtractorErrorCode.InvalidNavigationPropertyReference]: {
    description: "Invalid Navigation Property Reference",
    suggestion: "Ensure the navigation property points to a valid model field",
  },
  [ExtractorErrorCode.MissingNavigationPropertyReference]: {
    description: "Missing Navigation Property Reference",
    suggestion:
      "Navigation properties require a foreign key model attribute reference",
  },
  [ExtractorErrorCode.MissingManyToManyUniqueId]: {
    description: "Missing unique id on Many to Many navigation property",
    suggestion:
      "Define a unique identifier field for the Many-to-Many relationship",
  },
  [ExtractorErrorCode.MissingPrimaryKey]: {
    description: "Missing primary key on a model",
    suggestion: "Add a primary key field to your model (e.g., `id: number`)",
  },
  [ExtractorErrorCode.MissingWranglerEnv]: {
    description: "Missing a wrangler environment definition in the project",
    suggestion: "Add a @WranglerEnv class in your project.",
  },
  [ExtractorErrorCode.TooManyWranglerEnvs]: {
    description: "Too many wrangler environments defined in the project",
    suggestion: "Consolidate or remove unused @WranglerEnv's",
  },
  [ExtractorErrorCode.TooManyMiddlewares]: {
    description: "Too many middleware classes defined in the project",
    suggestion: "Only one @Middleware class is allowed per project",
  },
  [ExtractorErrorCode.MissingMiddlewareMethod]: {
    description: "Middleware class must have a 'handle' method",
    suggestion: "Add a 'handle' method to your @Middleware class",
  },
};

export function getErrorInfo(code: ExtractorErrorCode) {
  return errorInfoMap[code];
}

export class ExtractorError {
  context?: string;
  snippet?: string;

  constructor(public code: ExtractorErrorCode) {}

  addContext(fn: (val: string | undefined) => string | undefined) {
    this.context = fn(this.context ?? "");
  }
}

export type HttpResult<T = unknown> = {
  ok: boolean;
  status: number;
  data?: T;
  message?: string;
};

export type CidlType =
  | "Void"
  | "Integer"
  | "Real"
  | "Text"
  | "Blob"
  | { Inject: string }
  | { Object: string }
  | { Nullable: CidlType }
  | { Array: CidlType }
  | { HttpResult: CidlType };

export function isNullableType(ty: CidlType): boolean {
  return typeof ty === "object" && ty !== null && "Nullable" in ty;
}

export enum HttpVerb {
  GET = "GET",
  POST = "POST",
  PUT = "PUT",
  PATCH = "PATCH",
  DELETE = "DELETE",
}

export interface NamedTypedValue {
  name: string;
  cidl_type: CidlType;
}

export interface ModelAttribute {
  value: NamedTypedValue;
  foreign_key_reference: string | null;
}

export interface ModelMethod {
  name: string;
  is_static: boolean;
  http_verb: HttpVerb;
  return_type: CidlType | null;
  parameters: NamedTypedValue[];
}

export type NavigationPropertyKind =
  | { OneToOne: { reference: string } }
  | { OneToMany: { reference: string } }
  | { ManyToMany: { unique_id: string } };

export interface NavigationProperty {
  var_name: string;
  model_name: string;
  kind: NavigationPropertyKind;
}

export function getNavigationPropertyCidlType(
  nav: NavigationProperty
): CidlType {
  return "OneToOne" in nav.kind
    ? { Object: nav.model_name }
    : { Array: { Object: nav.model_name } };
}

export interface Model {
  name: string;
  primary_key: NamedTypedValue;
  attributes: ModelAttribute[];
  navigation_properties: NavigationProperty[];
  methods: Record<string, ModelMethod>;
  data_sources: Record<string, DataSource>;
  source_path: string;
}

export interface PlainOldObject {
  name: string;
  attributes: NamedTypedValue[];
  source_path: string;
}

export interface CidlIncludeTree {
  [key: string]: CidlIncludeTree;
}

export interface DataSource {
  name: string;
  tree: CidlIncludeTree;
}

export interface WranglerEnv {
  name: string;
  source_path: string;
}

export interface MiddlewareMethod {
  name: string;
  parameters: NamedTypedValue[];
}

export interface Middleware {
  class_name: string;
  method: MiddlewareMethod;
  source_path: string;
}

export interface CloesceAst {
  version: string;
  project_name: string;
  language: "TypeScript";
  wrangler_env: WranglerEnv;
  models: Record<string, Model>;
  poos: Record<string, PlainOldObject>;
  middleware: Middleware | null;
}
