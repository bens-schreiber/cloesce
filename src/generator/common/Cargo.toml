[package]
name = "common"
version = "0.0.1"
edition = "2024"

[dependencies]
anyhow = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
serde_with = "3.14.1"
toml = { workspace = true }
