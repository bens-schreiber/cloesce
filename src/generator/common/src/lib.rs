pub mod builder;
pub mod err;

use std::collections::BTreeMap;
use std::path::PathBuf;

use err::GeneratorErrorKind;
use err::Result;
use serde::Deserialize;
use serde::Serialize;
use serde_with::MapPreventDuplicates;
use serde_with::serde_as;

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq, Hash)]
pub enum CidlType {
    /// No type
    Void,

    /// SQLite integer
    Integer,

    /// SQLite floating point number
    Real,

    /// SQLite string
    Text,

    /// SQLite large structured data
    Blob,

    /// A dependency injected instance, containing a type name.
    Inject(String),

    /// A model, or plain old object, containing the name of the class.
    Object(String),

    /// A part of a model or plain object, containing the name of the class.
    ///
    /// Only valid as a method argument.
    Partial(String),

    /// An array of any type
    Array(Box<CidlType>),

    /// A REST API response, which can contain any type or nothing.
    HttpResult(Box<CidlType>),

    /// A wrapper denoting the type within can be null.
    /// If the inner value is void, represents just null.
    Nullable(Box<CidlType>),
}

impl CidlType {
    /// Returns the inner part of an array if the type is an array
    pub fn unwrap_array(&self) -> Option<&CidlType> {
        match self {
            CidlType::Array(inner) => Some(inner),
            _ => None,
        }
    }

    /// Returns the root most CidlType, being any non Model/Array/Result.
    pub fn root_type(&self) -> &CidlType {
        match self {
            CidlType::Array(inner) => inner.root_type(),
            CidlType::HttpResult(inner) => inner.root_type(),
            CidlType::Nullable(inner) => inner.root_type(),
            t => t,
        }
    }

    pub fn is_nullable(&self) -> bool {
        matches!(self, CidlType::Nullable(_))
    }

    pub fn array(cidl_type: CidlType) -> CidlType {
        CidlType::Array(Box::new(cidl_type))
    }

    pub fn nullable(cidl_type: CidlType) -> CidlType {
        CidlType::Nullable(Box::new(cidl_type))
    }

    pub fn null() -> CidlType {
        CidlType::Nullable(Box::new(CidlType::Void))
    }
}

#[derive(Serialize, Deserialize, Debug, PartialEq)]
pub enum HttpVerb {
    GET,
    POST,
    PUT,
    PATCH,
    DELETE,
}

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Hash)]
pub struct NamedTypedValue {
    pub name: String,
    pub cidl_type: CidlType,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct ModelAttribute {
    pub value: NamedTypedValue,
    pub foreign_key_reference: Option<String>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct ModelMethod {
    pub name: String,
    pub is_static: bool,
    pub http_verb: HttpVerb,
    pub return_type: CidlType,
    pub parameters: Vec<NamedTypedValue>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct IncludeTree(pub BTreeMap<String, IncludeTree>);

#[derive(Serialize, Deserialize, Debug)]
pub struct DataSource {
    pub name: String,
    pub tree: IncludeTree,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub enum NavigationPropertyKind {
    OneToOne { reference: String },
    OneToMany { reference: String },
    ManyToMany { unique_id: String },
}

#[derive(Serialize, Deserialize, Debug)]
pub struct NavigationProperty {
    pub var_name: String,
    pub model_name: String,
    pub kind: NavigationPropertyKind,
}

#[serde_as]
#[derive(Serialize, Deserialize, Debug)]
pub struct Model {
    pub name: String,
    pub primary_key: NamedTypedValue,
    pub attributes: Vec<ModelAttribute>,
    pub navigation_properties: Vec<NavigationProperty>,

    #[serde_as(as = "MapPreventDuplicates<_, _>")]
    pub methods: BTreeMap<String, ModelMethod>,

    #[serde_as(as = "MapPreventDuplicates<_, _>")]
    pub data_sources: BTreeMap<String, DataSource>,
    pub source_path: PathBuf,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct PlainOldObject {
    pub name: String,
    pub attributes: Vec<NamedTypedValue>,
    pub source_path: PathBuf,
}

#[derive(Serialize, Deserialize)]
pub enum InputLanguage {
    TypeScript,
}

#[derive(Serialize, Deserialize)]
pub struct WranglerEnv {
    pub name: String,
    pub source_path: PathBuf,
}

#[serde_as]
#[derive(Serialize, Deserialize)]
pub struct CloesceAst {
    pub version: String,
    pub project_name: String,
    pub language: InputLanguage,
    pub wrangler_env: WranglerEnv,

    #[serde_as(as = "MapPreventDuplicates<_, _>")]
    pub models: BTreeMap<String, Model>,

    #[serde_as(as = "MapPreventDuplicates<_, _>")]
    pub poos: BTreeMap<String, PlainOldObject>,
}

impl CloesceAst {
    pub fn from_json(path: &std::path::Path) -> Result<CloesceAst> {
        let cidl_contents = std::fs::read_to_string(path).map_err(|e| {
            GeneratorErrorKind::InvalidInputFile
                .to_error()
                .with_context(e.to_string())
        })?;
        serde_json::from_str::<CloesceAst>(&cidl_contents).map_err(|e| {
            GeneratorErrorKind::InvalidInputFile
                .to_error()
                .with_context(e.to_string())
        })
    }

    fn is_valid_object_ref(&self, o: &str) -> bool {
        self.models.contains_key(o) || self.poos.contains_key(o)
    }

    /// Ensures all `CidlTypes` are logically correct for the area, essentially doing
    /// the first level of semantic analysis for the generator.
    ///
    /// Returns error on
    /// - Model attributes with invalid SQL types
    /// - Primary keys with invalid SQL types
    /// - Invalid Model or Method map K/V
    /// - Unknown navigation property model references
    /// - Unknown model references in method parameters
    /// - Invalid method parameter types
    pub fn validate_types(&self) -> Result<()> {
        let ensure_valid_sql_type = |model: &Model, value: &NamedTypedValue| {
            let inner = match &value.cidl_type {
                CidlType::Nullable(inner) if matches!(inner.as_ref(), CidlType::Void) => {
                    fail!(GeneratorErrorKind::NullSqlType)
                }
                CidlType::Nullable(inner) => inner.as_ref(),
                other => other,
            };

            ensure!(
                matches!(
                    inner,
                    CidlType::Integer | CidlType::Real | CidlType::Text | CidlType::Blob
                ),
                GeneratorErrorKind::InvalidSqlType,
                "{}.{}",
                model.name,
                value.name
            );

            Ok(())
        };

        for (model_name, model) in &self.models {
            // Validate record
            ensure!(
                *model_name == model.name,
                GeneratorErrorKind::InvalidMapping,
                "Model record key did not match it's model name? {} : {}",
                model_name,
                model.name
            );

            // Validate PK
            ensure_valid_sql_type(model, &model.primary_key)?;

            // Validate attributes
            for a in &model.attributes {
                ensure_valid_sql_type(model, &a.value)?;

                if let Some(fk_model) = &a.foreign_key_reference {
                    // Validate the fk's model exists
                    ensure!(
                        self.models.contains_key(fk_model.as_str()),
                        GeneratorErrorKind::UnknownObject,
                        "{}.{} => {}?",
                        model.name,
                        a.value.name,
                        fk_model
                    );
                }
            }

            // Validate navigation props
            for nav in &model.navigation_properties {
                ensure!(
                    self.models.contains_key(nav.model_name.as_str()),
                    GeneratorErrorKind::UnknownObject,
                    "{} => {}?",
                    model.name,
                    nav.model_name
                );
            }

            // Validate methods
            for (method_name, method) in &model.methods {
                // Validate record
                ensure!(
                    *method_name == method.name,
                    GeneratorErrorKind::InvalidMapping,
                    "Method record key did not match it's method name? {}: {}",
                    method_name,
                    method.name
                );

                // Validate return type
                match &method.return_type {
                    CidlType::Object(o) => {
                        ensure!(
                            self.is_valid_object_ref(o),
                            GeneratorErrorKind::UnknownObject,
                            "{}.{}",
                            model.name,
                            method.name
                        );
                    }
                    CidlType::Partial(_) => {
                        fail!(
                            GeneratorErrorKind::UnexpectedPartialReturn,
                            "{}.{}",
                            model.name,
                            method.name,
                        )
                    }
                    _ => {}
                }

                // Validate method params
                for param in &method.parameters {
                    let root_type = param.cidl_type.root_type();

                    match root_type {
                        CidlType::Void => {
                            fail!(
                                GeneratorErrorKind::UnexpectedVoid,
                                "{}.{}.{}",
                                model.name,
                                method.name,
                                param.name
                            )
                        }
                        CidlType::Object(o) | CidlType::Partial(o) => {
                            ensure!(
                                self.is_valid_object_ref(o),
                                GeneratorErrorKind::UnknownObject,
                                "{}.{}.{}",
                                model.name,
                                method.name,
                                param.name
                            );

                            // TODO: remove this
                            if method.http_verb == HttpVerb::GET {
                                fail!(
                                    GeneratorErrorKind::NotYetSupported,
                                    "GET Requests currently do not support model parameters {}.{}.{}",
                                    model.name,
                                    method.name,
                                    param.name
                                )
                            }
                        }
                        _ => {
                            // Ignore
                        }
                    }
                }
            }
        }

        Ok(())
    }
}
