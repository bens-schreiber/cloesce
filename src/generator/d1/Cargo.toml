[package]
name = "d1"
version = "0.0.1"
edition = "2024"

[dependencies]
anyhow = { workspace = true }
common = { path = "../common" }
sea-query = "0.32.7"


[dev-dependencies]
toml = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
insta = { workspace = true }
