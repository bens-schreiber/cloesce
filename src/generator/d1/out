- alias=Student model=Student transition=None
  - alias=Course model=Course transition=Some(ManyToMany { unique_id: "StudentsCourses" })
