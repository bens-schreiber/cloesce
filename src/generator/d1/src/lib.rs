use std::collections::{BTreeMap, HashMap, VecDeque};

use common::{CidlForeignKey, CidlSpec, CidlType, D1Database, Model, WranglerSpec};

use anyhow::{Result, anyhow};
use sea_query::{Alias, ColumnDef, ForeignKey, SqliteQueryBuilder, Table};

/// Topological sort via Kahns algorithm
/// Returns an error if there is a cycle, duplicate models, or an unknown FK.
fn topo_sort<'a>(models: &'a [Model]) -> Result<Vec<&'a Model>> {
    let mut name_to_model = HashMap::<&'a str, &'a Model>::new();
    let mut name_to_in_degree = BTreeMap::<&'a str, usize>::new();
    let mut graph = BTreeMap::<&'a str, Vec<&'a str>>::new();

    // Detect dups, populate reverse lookup
    for model in models {
        match name_to_model.entry(&model.name) {
            std::collections::hash_map::Entry::Vacant(vacant_entry) => vacant_entry.insert(model),
            std::collections::hash_map::Entry::Occupied(_) => {
                return Err(anyhow!("Duplicate model name: {}", model.name));
            }
        };
    }

    // Increment in-degree
    for model in models {
        graph.entry(&model.name).or_default();
        name_to_in_degree.entry(&model.name).or_insert(0);

        for attribute in &model.attributes {
            if let Some(fk) = &attribute.foreign_key {
                if !name_to_model.contains_key(fk.as_str()) {
                    return Err(anyhow!(
                        "Unknown foreign key on model {}: {}",
                        model.name,
                        fk.as_str()
                    ));
                }

                if attribute.value.nullable {
                    // Important distinction: nullable FK's do not constrain table creation order,
                    // and thus can be left out of our topo sort
                    continue;
                }

                graph.entry(fk.as_str()).or_default().push(&model.name);
                *name_to_in_degree.entry(&model.name).or_insert(0) += 1;
            }
        }
    }

    let mut queue = name_to_in_degree
        .iter()
        .filter_map(|(&name, &v)| (v == 0).then_some(name))
        .collect::<VecDeque<&str>>();

    let mut ordered = vec![];
    while let Some(model) = queue.pop_front() {
        // Unwrap: safe because graph population covers the entire set
        ordered.push(*name_to_model.get(model).unwrap());

        if let Some(adjs) = graph.get(model) {
            for adj in adjs {
                // Unwrap: safe because `name_to_model` halts on unknown FK's
                let in_degree = name_to_in_degree.get_mut(adj).unwrap();
                *in_degree -= 1;

                if *in_degree == 0 {
                    queue.push_back(adj);
                }
            }
        }
    }

    if ordered.len() != models.len() {
        let cyclic_models: Vec<&str> = name_to_in_degree
            .iter()
            .filter_map(|(&name, &deg)| (deg > 0).then_some(name))
            .collect();
        return Err(anyhow!(
            "Cycle detected involving the following models: {}",
            cyclic_models.join(", ")
        ));
    }

    Ok(ordered)
}

pub struct D1Generator {
    cidl: CidlSpec,
    wrangler: WranglerSpec,
}

impl D1Generator {
    pub fn new(cidl: CidlSpec, wrangler: WranglerSpec) -> Self {
        Self { cidl, wrangler }
    }

    /// Validates and updates the Wrangler spec so that D1 can be used during
    /// code generation.
    pub fn wrangler(&self) -> WranglerSpec {
        // Validate existing database configs, filling in missing values with a default
        let mut res = self.wrangler.clone();
        for (i, d1) in res.d1_databases.iter_mut().enumerate() {
            if d1.binding.is_none() {
                d1.binding = Some(format!("D1_DB_{i}"));
            }

            if d1.database_name.is_none() {
                d1.database_name = Some(format!("{}_d1_{i}", self.cidl.project_name));
            }

            if d1.database_id.is_none() {
                eprintln!(
                    "Warning: Database \"{}\" is missing an id. \n https://developers.cloudflare.com/d1/get-started/",
                    d1.database_name.as_ref().unwrap()
                )
            }
        }

        // Ensure a database exists (if there are even models), provide a default if not
        if !self.cidl.models.is_empty() && res.d1_databases.is_empty() {
            res.d1_databases.push(D1Database {
                binding: Some(String::from("D1_DB")),
                database_name: Some(String::from("default")),
                database_id: None,
            });

            eprintln!(
                "Database \"default\" is missing an id. \n https://developers.cloudflare.com/d1/get-started/"
            );
        }

        res
    }

    /// Creates a Sqlite database schema from the CIDL Model AST
    // Note: Model names, attributes do not need to be santized as SeaQuery
    // wraps them in quote literals.
    pub fn sqlite(&self) -> Result<String> {
        let mut res = Vec::<String>::default();
        let mut pk_name_lookup = HashMap::<&str, String>::new();

        let models = topo_sort(&self.cidl.models)?;

        for model in models {
            let mut table = Table::create();
            table.table(Alias::new(model.name.clone()));

            let mut pk_name: Option<String> = None;
            for attribute in model.attributes.iter() {
                if attribute.primary_key {
                    if let Some(pk_name) = &pk_name {
                        return Err(anyhow!(
                            "Duplicate primary keys {} {}",
                            pk_name,
                            attribute.value.name
                        ));
                    }

                    if attribute.value.nullable {
                        return Err(anyhow!("A primary key cannot be nullable."));
                    }

                    if attribute.foreign_key.is_some() {
                        // TODO: Revisit this, should this design be allowed?
                        return Err(anyhow!("A primary key cannot be a foreign key"));
                    }
                }

                if let Some(_) = &attribute.foreign_key {
                    match attribute.value.cidl_type {
                        CidlType::Integer | CidlType::Model(_) => {
                            // Allowed types for a foreign key
                        }
                        _ => {
                            return Err(anyhow!(
                                "A foreign key must be either a Model or Integer type ({}.{})",
                                model.name,
                                attribute.value.name
                            ));
                        }
                    }
                }

                let mut column = ColumnDef::new(Alias::new(attribute.value.name.clone()));

                if attribute.primary_key {
                    column.primary_key();
                    pk_name = Some(attribute.value.name.clone());
                } else if !attribute.value.nullable {
                    column.not_null();
                }

                if let Some(fk) = &attribute.foreign_key {
                    match fk {
                        CidlForeignKey::OneToOne(fk_model_name) => {
                            table.foreign_key(
                                ForeignKey::create()
                                    .name(format!("fk_{}_{}", model.name, fk_model_name))
                                    .from(
                                        Alias::new(model.name.clone()),
                                        Alias::new(format!("{}Id", fk_model_name)),
                                    )
                                    .to(
                                        Alias::new(fk_model_name.clone()),
                                        Alias::new(
                                            pk_name_lookup.get(fk_model_name.as_str()).unwrap(),
                                        ), // Dog.id
                                    )
                                    .on_update(sea_query::ForeignKeyAction::Cascade)
                                    .on_delete(sea_query::ForeignKeyAction::Restrict),
                            );
                        }
                        _ => todo!(),
                    }
                }

                match &attribute.value.cidl_type {
                    CidlType::Integer => column.integer(),
                    CidlType::Real => column.decimal(),
                    CidlType::Text => column.text(),
                    CidlType::Blob => column.blob(),
                    other => return Err(anyhow!("Invalid SQLite type {:?}", other)),
                };

                table.col(column);
            }

            match pk_name {
                Some(pk_name) => pk_name_lookup.insert(&model.name, pk_name.clone()),
                None => return Err(anyhow!("Model {} is missing a primary key.", model.name)),
            };

            res.push(format!("{};", table.to_string(SqliteQueryBuilder)));
        }

        Ok(res.join("\n"))
    }
}

#[cfg(test)]
mod tests {

    use itertools::Itertools;

    use crate::{D1Generator, topo_sort};
    use common::{
        Attribute, CidlForeignKey, CidlSpec, CidlType, InputLanguage, Model, TypedValue,
        WranglerSpec,
    };

    fn create_cidl(models: Vec<Model>) -> CidlSpec {
        CidlSpec {
            version: "1.0".to_string(),
            project_name: "test".to_string(),
            language: InputLanguage::TypeScript,
            models,
        }
    }

    fn create_wrangler() -> WranglerSpec {
        WranglerSpec {
            d1_databases: vec![],
        }
    }

    fn create_model(name: &str, fk: Vec<&str>) -> Model {
        let mut attributes = fk
            .iter()
            .map(|fk| Attribute {
                value: TypedValue {
                    name: format!("{}Id", fk),
                    cidl_type: CidlType::Integer,
                    nullable: false,
                },
                foreign_key: Some(CidlForeignKey::OneToOne(fk.to_string())),
                primary_key: false,
            })
            .collect::<Vec<Attribute>>();

        attributes.push(Attribute {
            value: TypedValue {
                name: "id".to_string(),
                cidl_type: CidlType::Integer,
                nullable: false,
            },
            foreign_key: None,
            primary_key: true,
        });

        Model {
            name: name.to_string(),
            attributes,
            methods: vec![],
            data_sources: vec![],
            source_path: "".into(),
        }
    }

    fn is_topo_ordered(sorted: &[&Model]) -> bool {
        use std::collections::HashSet;

        let mut visited = HashSet::<String>::new();

        for &model in sorted {
            for attribute in &model.attributes {
                if let Some(fk) = &attribute.foreign_key {
                    if !visited.contains(fk.as_str()) && !attribute.value.nullable {
                        return false;
                    }
                }
            }

            visited.insert(model.name.clone());
        }

        true
    }

    #[test]
    fn test_primary_key_and_value_yields_sqlite() {
        // Arrange
        let spec = create_cidl(vec![Model {
            source_path: "./models/user.cloesce.ts".into(),
            name: String::from("User"),
            attributes: vec![
                Attribute {
                    value: TypedValue {
                        name: String::from("id"),
                        cidl_type: CidlType::Integer,
                        nullable: false,
                    },
                    primary_key: true,
                    foreign_key: None,
                },
                Attribute {
                    value: TypedValue {
                        name: String::from("name"),
                        cidl_type: CidlType::Text,
                        nullable: true,
                    },
                    primary_key: false,
                    foreign_key: None,
                },
                Attribute {
                    value: TypedValue {
                        name: String::from("age"),
                        cidl_type: CidlType::Integer,
                        nullable: false,
                    },
                    primary_key: false,
                    foreign_key: None,
                },
            ],
            methods: vec![],
            data_sources: vec![],
        }]);

        let d1gen = D1Generator::new(spec, create_wrangler());

        // Act
        let sql = d1gen.sqlite().expect("gen_sqlite to work");

        // Assert
        assert!(sql.contains("CREATE TABLE"));
        assert!(sql.contains("\"id\" integer PRIMARY KEY"));
        assert!(sql.contains("\"name\" text"));
        assert!(sql.contains("\"age\" integer NOT NULL"));
    }

    #[test]
    fn test_duplicate_primary_key_error() {
        // Arrange
        let spec = create_cidl(vec![Model {
            source_path: "./models/user.cloesce.ts".into(),
            name: String::from("User"),
            attributes: vec![
                Attribute {
                    value: TypedValue {
                        name: String::from("id"),
                        cidl_type: CidlType::Integer,
                        nullable: false,
                    },
                    primary_key: true,
                    foreign_key: None,
                },
                Attribute {
                    value: TypedValue {
                        name: String::from("user_id"),
                        cidl_type: CidlType::Integer,
                        nullable: false,
                    },
                    primary_key: true,
                    foreign_key: None,
                },
            ],
            methods: vec![],
            data_sources: vec![],
        }]);

        let d1gen = D1Generator::new(spec, create_wrangler());

        // Act
        let err = d1gen.sqlite().unwrap_err();

        // Assert
        assert!(err.to_string().contains("Duplicate primary keys"));
    }

    #[test]
    fn test_nullable_primary_key_error() {
        // Arrange
        let spec = create_cidl(vec![Model {
            source_path: "./models/user.cloesce.ts".into(),
            name: String::from("User"),
            attributes: vec![Attribute {
                value: TypedValue {
                    name: String::from("id"),
                    cidl_type: CidlType::Integer,
                    nullable: true,
                },
                primary_key: true,
                foreign_key: None,
            }],
            methods: vec![],
            data_sources: vec![],
        }]);

        let d1gen = D1Generator::new(spec, create_wrangler());

        // Act
        let err = d1gen.sqlite().unwrap_err();

        // Assert
        assert!(
            err.to_string()
                .contains("A primary key cannot be nullable.")
        );
    }

    #[test]
    fn test_missing_primary_key_error() {
        // Arrange
        let spec = create_cidl(vec![Model {
            source_path: "./models/user.cloesce.ts".into(),
            name: String::from("User"),
            attributes: vec![Attribute {
                value: TypedValue {
                    name: String::from("id"),
                    cidl_type: CidlType::Integer,
                    nullable: true,
                },
                primary_key: false,
                foreign_key: None,
            }],
            methods: vec![],
            data_sources: vec![],
        }]);

        let d1gen = D1Generator::new(spec, create_wrangler());

        // Act
        let err = d1gen.sqlite().unwrap_err();

        // Assert
        assert!(err.to_string().contains("User is missing a primary key."));
    }

    #[test]
    fn test_topo_sort_yields_correct_order() {
        // Arrange
        // note: this test is kind of ridiculous
        let creators: Vec<Box<dyn Fn() -> Model>> = vec![
            Box::new(|| create_model("Treat", vec![])),
            Box::new(|| create_model("Food", vec![])),
            Box::new(|| create_model("Dog", vec!["Treat", "Food"])),
            Box::new(|| create_model("Independent", vec![])),
            Box::new(|| create_model("Person", vec!["Dog"])),
        ];

        // 5 items, 120 permutations
        for permutation in creators.iter().permutations(creators.len()) {
            // Act
            let perm_slice: Vec<Model> = permutation.into_iter().map(|f| f()).collect();

            let sorted = topo_sort(&perm_slice).expect("topo_sort failed");

            // Assert
            assert!(is_topo_ordered(&sorted));
        }
    }

    #[test]
    fn test_duplicate_model_error() {
        // Arrange
        let models = vec![create_model("User", vec![]), create_model("User", vec![])];

        // Act
        let err = topo_sort(&models).unwrap_err();

        // Assert
        assert!(err.to_string().contains("Duplicate model name"));
    }

    #[test]
    fn test_unknown_foreign_key_error() {
        // Arrange
        let models = vec![create_model("User", vec!["NonExistent"])];

        // Act
        let err = topo_sort(&models).unwrap_err();

        // Assert
        assert!(err.to_string().contains("Unknown foreign key"));
    }

    #[test]
    fn test_cycle_detection_error() {
        // Arrange
        // A -> B -> C -> A
        let models = vec![
            create_model("A", vec!["B"]),
            create_model("B", vec!["C"]),
            create_model("C", vec!["A"]),
        ];

        // Act
        let err = topo_sort(&models).unwrap_err();

        // Assert
        assert!(err.to_string().contains("Cycle detected"));
    }

    #[test]
    fn test_nullability_prevents_cycle_error() {
        // Arrange
        // A -> B -> C -> Nullable<A>
        let mut models = vec![
            create_model("A", vec!["B"]),
            create_model("B", vec!["C"]),
            create_model("C", vec!["A"]),
        ];
        models[2].attributes[0].value.nullable = true;

        // Act
        let sorted = topo_sort(&models);

        // Assert
        assert!(is_topo_ordered(&sorted.unwrap()));
    }

    #[test]
    fn test_one_to_one_fk_yields_sqlite() {
        // Arrange
        let spec = create_cidl(vec![
            create_model("User", vec!["Dog"]),
            create_model("Dog", vec![]),
        ]);
        let d1gen = D1Generator::new(spec, create_wrangler());

        // Act
        let sql = d1gen.sqlite().expect("gen_sqlite to work");

        // Assert
        assert!(sql.contains(
            r#"FOREIGN KEY ("DogId") REFERENCES "Dog" ("id") ON DELETE RESTRICT ON UPDATE CASCADE "#
        ));
    }
}
