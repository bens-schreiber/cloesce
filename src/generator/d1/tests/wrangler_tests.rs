use common::{CloesceAst, wrangler::WranglerFormat};
use d1::D1Generator;

use anyhow::Result;
use insta::assert_snapshot;

use std::path::PathBuf;

#[test]
fn test_serialize_wrangler_spec() {
    // Filled TOML
    {
        let wrangler_path = PathBuf::from("../../test_fixtures/wrangler.toml");
        WranglerFormat::from_path(&wrangler_path).expect("Wrangler file to serialize");
    }

    // Empty TOML
    {
        WranglerFormat::Toml(toml::from_str("").unwrap())
            .as_spec()
            .expect("Wrangler file to serialize");
    }

    // Filled JSON
    {
        let wrangler_path = PathBuf::from("../../test_fixtures/wrangler.json");
        WranglerFormat::from_path(&wrangler_path).expect("Wrangler file to serialize");
    }

    // Empty JSON
    {
        WranglerFormat::Json(serde_json::from_str("{}").unwrap())
            .as_spec()
            .expect("Wrangler file to serialize");
    }
}

#[test]
fn test_generate_empty_wrangler_snap() -> Result<()> {
    // Arrange
    let ast = {
        let cidl_path = PathBuf::from("../../test_fixtures/cidl.json");
        let cidl_contents = std::fs::read_to_string(cidl_path)?;
        serde_json::from_str::<CloesceAst>(&cidl_contents)?
    };

    let wrangler = WranglerFormat::Toml(toml::from_str("").unwrap());

    let d1gen = D1Generator::new(ast, wrangler.as_spec()?);

    // Act
    let updated_wrangler = d1gen.wrangler();

    // Assert
    assert_snapshot!(
        "updated_wrangler_from_empty",
        format!("{}", toml::to_string_pretty(&updated_wrangler).unwrap())
    );

    Ok(())
}
