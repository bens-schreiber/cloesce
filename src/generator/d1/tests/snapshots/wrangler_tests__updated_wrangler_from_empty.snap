---
source: d1/tests/wrangler_tests.rs
expression: "format!(\"{}\", toml::to_string_pretty(&updated_wrangler).unwrap())"
---
[[d1_databases]]
binding = "D1_DB"
database_name = "default"
